File: core/src/main/java/org/robotframework/roc/core/services/EnvironmentService.java
package org.robotframework.roc.core.services;

import org.robotframework.roc.core.dto.EnvironmentCreateDto;
import org.robotframework.roc.core.dto.EnvironmentUpdateDto;
import org.robotframework.roc.core.models.Environment;

import java.util.List;
import java.util.Optional;

public interface EnvironmentService extends CRUDService<Long> {
    Environment createEnvironment(Environment environment);

    Environment createEnvironment(Long id, EnvironmentCreateDto dto) throws Exception;

    Environment updateEnvironment(Long id, EnvironmentUpdateDto dto) throws Exception;

    Environment updateEnvironment(Environment environment, EnvironmentUpdateDto dto) throws Exception;

    void deleteEnvironment(Long id);

    List<Environment> getEnvironments(Long projectId);

    Optional<Environment> getEnvironmentById(Long id);

}

File: packages/platform/src/environment/environment.service.ts
import { Inject, Injectable } from '@nestjs/common';
import {
  Environment,
  EnvironmentCreateDto,
  EnvironmentRetrieveDto,
  EnvironmentUpdateDto,
  Id,
} from '@roc/core';
import { EnvironmentRedisRepository } from './environment.repository';

@Injectable()
export class EnvironmentService {
  @Inject()
  private readonly repository: EnvironmentRedisRepository;

  public getRepository() {
    return this.repository;
  }

  public async existsWithInProject(
    projectId: Id,
    environmentId: Id,
  ): Promise<boolean> {
    return this.repository.existsById(`${projectId}.${environmentId}`);
  }

  public async findAllByProjectId(
    projectId: Id,
  ): Promise<EnvironmentRetrieveDto[]> {
    return EnvironmentRetrieveDto.fromMany(
      await this.repository.findByKey(`${projectId}.*`),
    );
  }

  public async getEnvironmentById(
    projectId: Id,
    id: Id,
    asDto = true,
  ): Promise<EnvironmentRetrieveDto | Environment> {
    if (asDto) {
      return EnvironmentRetrieveDto.from(
        await this.repository.getOneById(`${projectId}.${id}`),
      );
    } else {
      await this.repository.getOneById(`${projectId}.${id}`);
    }
  }

  public async createNewEnvironment(
    projectId: Id,
    environmentCreateDto: EnvironmentCreateDto,
  ): Promise<Environment> {
    const environment = Environment.fromPlainObject(environmentCreateDto);
    environment.projectId = projectId;
    return this.repository.save(environment);
  }

  public async deleteEnvironment(projectId: Id, id: Id) {
    return this.repository.deleteById(`${projectId}.${id}`);
  }

  public async updateEnviromentByIds(
    projectId: Id,
    environmentId: Id,
    dto: EnvironmentUpdateDto,
  ) {
    const env: Environment = await this.repository.getOneById(
      `${projectId}.${environmentId}`,
    );

    env.description = dto.description;
    env.tags = dto.tags;
    env.variables = dto.variables;

    this.repository.save(env);
    return env;
  }
}
